src/servers/http_server.h

```diff
@@ -204,6 +203,9 @@ class HTTPAPIServer : public HTTPServer {
     // lifetime of the request.
     std::list<std::vector<char>> serialized_data_;
 
+    std::atomic<uint32_t> complete_count_;
```

Comment:
Why does the request lifetime management get more complicated?

Reply:
The HTTP request lifecycle wasn't being managed correctly, before it assumes the requests can be released when response complete callback is called, which is not the case because the backend can release the request after response is sent, so I think the lifecycle should be done when both the response and request are done.

In this particular case, Triton core will call "response complete -> request complete (log timestamp to `Trace`)", so with the previous HTTP request lifecycle, the trace will operate on released `Trace` object.

It was working before because the frontend and the Triton trace report timestamps to `TraceManager` directly and there is no dependency between the Triton trace and HTTP request, whereas now they both report to `Trace` which is owned by the HTTP request.

Reply:
In the old lifecycle was the HTTP request needed to live until the response completed? Or is that a new requirement due to the new way of handling Trace? I'm worried about introducing a memory leak... the existing HTTP endpoint is quite convoluted and makes these kind of changes risky.

Reply:
> In the old lifecycle was the HTTP request needed to live until the response completed?

Yes, the old way requires HTTP request lives until the response completed. The new way extends the requirement to that HTTP request lives until both response and request are completed.

I can make the `TRITONSERVER_InferenceTrace` to carry the pointer to a `shared_ptr<Trace>` and delete it when Triton trace finish its part. Then HTTP request lifecycle won't need to be changed, even if it is released before Triton trace, the `Trace` object will still be valid.

Reply:
So the old way is ok as long as we don't actually need the HTTP request object on request completion. Is that the case?
If so, I would prefer to use shared_ptr<Trace> idea and not change the HTTP request object lifecycle handling.